sol/main.c

```diff
@@ -0,0 +1,449 @@
+#define REQ_QUEUE_BW_RATE (.05)
+
+/* Token bucket rate approximation error. */
+#define RTE_SCHED_TB_RATE_CONFIG_ERR (1e-7)
+
+/*
+ * Scaling for cycles_per_byte calculation; chosen so that minimum rate
+ * is 480 bit/sec. Adapted for use from the DPDK scheduling library.
+ */
+#define RTE_SCHED_TIME_SHIFT (8)
+
+/*
+ * Solicitor configuration allocated by alloc_sol_conf(), needed
+ * by gk_solicitor_enqueue() since the GK instances do not have
+ * any visibility into the Solicitor.
+ */
+static struct sol_config *sol_conf;
+
+static void
+insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+	uint8_t priority)
+{
+	uint8_t next;
+
+	req_queue->priorities[priority] = pr;
+
+	/* This is the first packet in the queue. */
+	if (req_queue->len == 0) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		req_queue->lowest_priority = priority;
+		return;
+	}
+
+	/* Not the first packet, but still insert at the head of the queue. */
+	if (priority > req_queue->highest_priority) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		return;
+	}
+
+	/* Insert in middle or end of queue. */
+	if (priority < req_queue->lowest_priority)
+		req_queue->lowest_priority = priority;
+	for (next = priority + 1; next < req_queue->num_priorities; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
+		}
+	}
+
+	rte_panic("sol: %s failed to insert a request of a new priority\n",
+		__func__);
+}
+
+static void
+drop_lowest_priority_pkt(struct sol_config *sol_conf,
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr =
+		req_queue->priorities[req_queue->lowest_priority];
+	struct rte_mbuf *lowest_pr_pkt = lowest_pr->pkt;
+	struct priority_req *next_lowest_pr;
+
+	/* Only one packet in the queue. */
+	if (unlikely(req_queue->len == 1)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->lowest_priority = req_queue->num_priorities;
+		goto drop;
+	}
+
+	next_lowest_pr = list_last_entry(&lowest_pr->list,
+		struct priority_req, list);
+
+	/* The lowest priority packet was the only one of that priority. */
+	if (lowest_pr->priority != next_lowest_pr->priority) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->lowest_priority = next_lowest_pr->priority;
+		goto drop;
+	}
+
+	req_queue->priorities[lowest_pr->priority] = next_lowest_pr;
+
+drop:
+	list_del(&lowest_pr->list);
+	mb_free_entry(&sol_conf->mb, lowest_pr);
+	rte_pktmbuf_free(lowest_pr_pkt);
+	req_queue->len--;
+}
+
+static void
+enqueue_req(struct sol_config *sol_conf, struct priority_req *req)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	uint8_t priority = req->priority;
+
+	if (req_queue->len == sol_conf->pri_req_max_len) {
+		/* New packet is lowest priority, so drop it. */
+		if (req_queue->lowest_priority == priority) {
```

Comment:
Shouldn't it be ``req_queue->lowest_priority >= priority``? Since if the priority is smaller than the lowest one, it will be dropped as well.